### PR TITLE
Chapter 2

### PR DESCRIPTION
**Pull request by mridul2899**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -336,8 +336,13 @@ from it!
 ghci> :l src/Chapter2.hs
 -}
 subList :: Int -> Int -> [a] -> [a]
-subList = error "subList: Not implemented!"
-
+subList left right list
+    | left >= 0 && right >= 0 =
```
Comment (chshersh) [inline src/Chapter2.hs:340]: Your implementation is correct when the right bound is lower than the left. But I think you can implement it more efficiently by adding a separate case to your check :slightly_smiling_face:

File: src/Chapter2.hs [hunk 2]
```diff
@@ -621,7 +634,10 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven [] = []
+takeEven (x1:_:xs) = x1 : takeEven xs
+takeEven (x1:xs) = x1 : takeEven xs
```
Comment (chshersh) [inline src/Chapter2.hs:640]: Here actually only one pattern left: singleton list. So you can write it explicitly to be sure that you've covered all cases :slightly_smiling_face: 
```suggestion
takeEven [x] = [x]
```

File: src/Chapter2.hs [hunk 3]
```diff
@@ -741,9 +756,8 @@ the list with only those lists that contain a passed element.
 
 🕯 HINT: Use the 'elem' function to check whether an element belongs to a list
 -}
-contains = error "contains: Not implemented!"
-
-
+contains :: Eq a => a -> [[a]] -> [[a]]
```
Comment (chshersh) [inline src/Chapter2.hs:759]: Nice work on specifying the polymorphic type signature here! :clap:

File: src/Chapter2.hs [hunk 4]
```diff
@@ -842,7 +858,10 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate n l
+    | n < 0 = []
+    | otherwise = take (length l) (drop n (cycle l))
```
Comment (chshersh) [inline src/Chapter2.hs:864]: That's a correct solution! Though, it does a little bit more than work when the given number is big and the list is small. You will cycle through the list a lot. Maybe you can think of a way to optimize this :slightly_smiling_face:

File: src/Chapter2.hs [hunk 5]
```diff
@@ -858,7 +877,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind (x:xs) = rewind xs ++ [x]
```
Comment (chshersh) [inline src/Chapter2.hs:882]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by chshersh — APPROVED** [review]
Great work on Chapter Two tasks!
And well done with the advanced task :100: 

I left some comments and suggestions regarding possible code improvements :slightly_smiling_face:

**Comment by mridul2899** [comment]
Thank you very much for your suggestions. They were quite valuable. I have made the appropriate changes to ensure that implementations are now efficient.

**Comment by vrom911** [comment]
The solutions look awesome 🌟 
 I'm going to close this PR, because we can't merge the solutions to our repo :slightly_smiling_face: 
Feel free to open PRs with solutions to other chapters!

**Pull request closed without merging**